lib/iris/fileformats/netcdf.py

```diff
@@ -1593,57 +1725,158 @@ def _get_dim_names(self, cube):
+                        # coord, not an identical one (i.e. "is" not "==")
+                        name = self._dim_names_and_coords.name(mesh_coord)
+                        stored_coord = self._dim_names_and_coords.coord(name)
+                        match = mesh_coord is stored_coord
+                    if match:
+                        # Use the previous name for this dim of this mesh, from
+                        # a previously saved cube with the same mesh.
+                        dim_name = self._dim_names_and_coords.name(mesh_coord)
+                    else:
+                        if location == "node":
+                            # always 1-d
+                            (dim_length,) = mesh_coord.shape
+                        else:
+                            # extract source dim, respecting dim-ordering
+                            dim_length = mesh_coord.shape[mesh_coord.src_dim]
+                        location_dim_attr = f"{location}_dimension"
+                        dim_name = getattr(mesh, location_dim_attr)
+                        if dim_name is None:
+                            dim_name = f"{mesh.name()}_{location}"
+                        while dim_name in self._existing_dim:
+                            dim_name = self._increment_name(dim_name)
+
+                        record_dimension(dim_name, dim_length, dim_coords)
 
+                    # Store the mesh dims indexed by location
+                    mesh_location_dimnames[location] = dim_name
+
+            # Identify the cube dimension which maps to the mesh.
+            (cube_mesh_dim,) = cube.coords(mesh_coords=True)[0].cube_dims(cube)
+            # Record the actual file dimension names for each mesh.
+            self._mesh_dims[mesh] = mesh_location_dimnames.copy()
+
+        mesh_dimensions = dimension_names.copy()
+
+        # Get the cube dimensions, in order.
+        dimension_names.clear()
+        for dim in range(cube.ndim):
+            if dim == cube_mesh_dim:
+                # Handle a mesh dimension: we already named this.
+                dim_coords = []
+                dim_name = self._mesh_dims[mesh][cube.location]
             else:
-                # No CF-netCDF coordinates describe this data dimension.
-                dim_name = "dim%d" % dim
-                if dim_name in self._existing_dim:
-                    # Increment name if conflicted with one already existing.
-                    if self._existing_dim[dim_name] != cube.shape[dim]:
+                # Get a name from the dim-coord (if any).
+                dim_coords = cube.coords(dimensions=dim, dim_coords=True)
+                if dim_coords:
+                    # Derive a dim name from a coord.
+                    coord = dim_coords[0]  # always have at least one
+
+                    # Add only dimensions that have not already been added.
+                    if coord in self._dim_names_and_coords.coords:
```

Comment:
@pp-mo We could refine this behaviour (as discussed) here by the `coord` and `name` methods on the `CFNameCoordMap` returning `None` rather than raising a `KeyError`.

That change is all self contained within this function (I think)... I'll leave it for your to consider

Reply:
:+1:
as discussed offline, quite possible + looks like it gives a simpler approach
TBD